Fix #2225: Make library carousel title links more obvious and remove "View All" button

abhijit5893 opened the pull request:
Recreated the[ pull request](https://github.com/oppia/oppia/pull/2505) to avoid merge conflicts. @seanlip  Please take a look.

seanlip commented:
@abhijit5893, this looks great, thanks! The code looks good, and I have no additional comments on that front.

I do have one UI request, though. When hovering over the space between the cards, the right arrow next to the header disappears. This is a bit jarring, especially as I move my mouse from left to right along a row of cards. Would it be possible to keep the right arrow there as long as the mouse is somewhere within the group's "row" (but not necessarily over a card)?

Thanks!

abhijit5893 commented:
@seanlip Yes. That would be possible. I can try applying the same behavior to the entire carousel div. That would also mean that If you hover above the card (within the margin), it would create the arrow.

seanlip commented:
That would feel slightly weird, I think, because there'd be an invisible breakpoint for no good reason (from the user's perspective). Is it possible for that not to happen?

abhijit5893 commented:
Not sure. Let me check.

abhijit5893 commented:
@seanlip I have tried different approaches but without success. Please let me know how to proceed.

seanlip commented:
One way around this might be to put an additional div around the carousel, and have that receive the on-mouseover event instead. Would that be possible?

seanlip commented:
Hi @abhijit5893, I talked to @BenHenning about this in some detail.

Here's our suggestion: could you please have the mouseover area be a rectangle that includes the group header + the carousel, instead? Ideally, its bottom edge would match the bottom edge of the cards, but if the cards have a margin it's fine if it juts out a small bit. That would be much better than having the mouse cursor change when moving between the individual cards.

This screenshot shows roughly where an ideal mouseover area would be:

![example](https://cloud.githubusercontent.com/assets/10575562/19261340/50528c0a-8f45-11e6-8da1-19a72ab2cd39.png)

Would this be possible? Thanks!

abhijit5893 commented:
@seanlip  Yes. I will push the changes by the end of today.

abhijit5893 commented:
@seanlip @BenHenning Please take a look

seanlip commented:
Hi @abhijit5893 -- please note the boxed outline in my [earlier comment](https://github.com/oppia/oppia/pull/2545#issuecomment-252829203), which includes the header. The current UI doesn't do so; it only covers the carousel. Could you adjust things so that the mouseover area matches that outline? Thanks!

abhijit5893 commented:
@seanlip Please take a look.

seanlip commented:
Hi @abhijit5893, this looks good on all counts -- both code and UI! I have no objections. I just had one small comment.

However, I want to ask @BenHenning to poke around at the UI on this, because this design was his idea originally and I want to make sure it matches the original intention. In particular, @BenHenning, two specific questions I have for you are:
- Is the caret that appears next to the library group header part of the header (i.e., when you hover over it, should the cursor change into a pointer)?
- Same question, but for the whitespace on the right of the caret.

From playing with the UI, I feel that at least the first one should be part of the link, but am not sure about the second, so would appreciate clarification.

Thanks!

BenHenning commented:
> However, I want to ask @BenHenning to poke around at the UI on this, because this design was his idea originally and I want to make sure it matches the original intention. In particular, @BenHenning, two specific questions I have for you are:
>
* Is the caret that appears next to the library group header part of the header (i.e., when you hover over it, should the cursor change into a pointer)?
* Same question, but for the whitespace on the right of the caret.

Yes to the first, no to the second. Similar to how we use a bounding box for the entire carousel and its header to know when to show the caret, we should have something similar around the header and the caret. If the mouse is over the caret, in the whitespace left of the caret and right of the header, or over the entire header then the header and caret should be highlighted.

Other feedback:
* The header should be highlighted when hovered over, rather than underlined
* The caret should be larger (it should be about as tall as the lowercase letters)

For a reference on caret size, see the pictures in this comment: https://github.com/oppia/oppia/issues/1839#issuecomment-222596560

seanlip commented:
Thanks, Ben!

Just a thought: an alternative to achieve the desired hover effect might be to just make the caret a part of the header anchor?

Also, @BenHenning what do you mean by "highlighted" -- do you mean that it changes color? If so, what should it change color to? /cc @markhalpin

BenHenning commented:
> Just a thought: an alternative to achieve the desired hover effect might be to just make the caret a part of the header anchor?
Sure, if we can make it appear/disappear but still be part of the anchor.

> Also, @BenHenning what do you mean by "highlighted" -- do you mean that it changes color? If so, what should it change color to? /cc @markhalpin
The picture in the #1839 comment shows the difference between the standard header color (dark) and the highlighted color (lighter) for Netflix. We have a light background, so I'm not sure what the highlight color should be. I'm merely suggesting we have one. We could maybe darken the current headers and then have a lighter highlight; what do you think @markhalpin?

markhalpin commented:
I think let's go with leaving the current color but switching it to the color in "Begin Your Adventure..." to highlight. It's not a super-dramatic change, but it's noticeable, and it'll work with the background. See attached.
![library_highlight](https://cloud.githubusercontent.com/assets/13863393/19485116/12e6d400-9528-11e6-810f-bd72aa900843.jpg)

seanlip commented:
Thanks, @markhalpin!

@abhijit5893 -- I think this should give enough info for a fix. Would it be possible to make the updates, please? Then this can go in.

Thanks!

BenHenning commented:
Color change sounds great to me. Nice suggestion.

On Tue, Oct 18, 2016 at 1:03 PM, Sean Lip <notifications@github.com> wrote:

> Thanks, @markhalpin <https://github.com/markhalpin>!
>
> @abhijit5893 <https://github.com/abhijit5893> -- I think this should give
> enough info for a fix. Would it be possible to make the updates, please?
> Then this can go in.
>
> Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/2545#issuecomment-254622767>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMYdvpSy4ZtqCra-hFhDAB7ZFGSmK1I2ks5q1SYQgaJpZM4KMJte>
> .
>

abhijit5893 commented:
@seanlip  Please take a look

seanlip commented:
@abhijit5893 did you test this? I see a couple of serious issues:
- Clicking on the header does not do anything.
- The cards jiggle when the group is moused over, which is really disconcerting...

The easiest way to test is by reloading all explorations from the /admin page and visiting /library. Please ensure that the UI/UX works properly before sending in the next review request!

abhijit5893 commented:
I did test it. I will make a video just to make sure.

seanlip commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -90,20 +90,17 @@ <h2 class="oppia-library-h2" translate="I18N_LIBRARY_SUB_HEADER"></h2>
-                </a>
+                </span>
+                <span ng-show="activeGroupIndex === $index"><i class="material-icons md-40"
+                  style="height:100%;">&#xE315;</i></span>
```
Why do you need the style here? I suspect that this is what is causing the "jiggling" issues.

abhijit5893 commented:
@seanlip  The flickering or jiggling effect was resolved when I changed ng-mousein and ng-mouseout to ng-mouseenter and ng-mouseleave. I have also removed the style which was not necessary. Please let me know if it functions as expected.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -1027,6 +1027,17 @@ textarea {
   min-width: 126px;
 }
 
+.oppia-library-group-header-active{
```
@abhijit5893, did you see this comment?

abhijit5893 replied:
Yes. This was fixed.

seanlip replied:
Not according to this page: https://github.com/oppia/oppia/pull/2545/files

Also, I noticed you're rebasing? You don't need to do that -- feel free to just submit incremental commits. They'll all just get squashed-on-merge anyway, and it's easier to see what has changed from one commit to the next in the interim.

seanlip commented:
@abhijit5893 There's still a jiggle on my screen (although a smaller one than there was before). In addition, the header is still not a link...

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -1027,6 +1030,17 @@ textarea {
   min-width: 126px;
 }
 
+.oppia-library-group-header-active {
```
I thought, in a previous comment, we talked about this being an additive class instead of one that replaces .oppia-library-group-header? That would reduce duplication -- see previous comment for more details.

abhijit5893 replied:
Ok that means that additive class just changes color (to avoid duplication of other css properties). Gotcha.

seanlip commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -90,20 +90,16 @@ <h2 class="oppia-library-h2" translate="I18N_LIBRARY_SUB_HEADER"></h2>
-                   translate="<[group.header_i18n_id]>">
-                </a>
+              <h2 ng-class="{'oppia-library-group-header-active': activeGroupIndex === $index, 'oppia-library-group-header': activeGroupIndex != $index }">
+                <span ng-if="group.has_full_results_page" translate="<[group.header_i18n_id]>"
```
This needs to be an actual link ('a' tag) for accessibility reasons.

abhijit5893 replied:
If you add the a tag. Then the hyperlink will be underlined. Is it ok?

seanlip replied:
No, it should still look like the mocks. But you can style using CSS to remove the underline (see the text-decoration attr).

abhijit5893 replied:
Got it. Thanks.

seanlip commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -92,10 +92,11 @@ <h2 class="oppia-library-h2" translate="I18N_LIBRARY_SUB_HEADER"></h2>
-                <span ng-if="group.has_full_results_page" translate="<[group.header_i18n_id]>"
-                  ng-click="showFullResultsPage(group.categories, group.full_results_url)">
-                </span>
+              <h2 ng-class="{'oppia-library-group-header active': activeGroupIndex === $index, 'oppia-library-group-header': activeGroupIndex != $index }">
```
Instead of this, do
```
<h2 class="oppia-library-group-header" ng-class="{'active': activeGroupIndex === $index}">

seanlip commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -92,10 +92,11 @@ <h2 class="oppia-library-h2" translate="I18N_LIBRARY_SUB_HEADER"></h2>
+                   ng-click="showFullResultsPage(group.categories, group.full_results_url)"
+                   translate="<[group.header_i18n_id]>">
+                </a>
                 <span ng-show="activeGroupIndex === $index"><i class="material-icons md-32">&#xE315;</i></span>
```
Should this be included in the anchor tag?

seanlip commented:
UI is much better now, thanks! Just that the full header (including the >) needs to act as a link, not just the text. But the jiggling seems to be fixed.

Also, I had two code comments.

Can you also do a quick video of the final behaviour so that others can take a look?

Thanks!

abhijit5893 commented:
Here is a [link](https://youtu.be/BCJ6cO-yWHA) to the demo video. Please take a look. @seanlip

seanlip commented:
Hi @abhijit5893, this looks great, and we can merge it when the Travis checks pass. Congrats on getting it done, and thanks!
